modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/MessageBoardMessageResourceImpl.java

```diff
@@ -191,6 +203,18 @@ public Rating getMessageBoardMessageMyRating(Long messageBoardMessageId)
 			filter, pagination, sorts);
 	}
 
+	@Override
+	public MessageBoardMessage
+			getSiteMessageBoardMessageByExternalReferenceCode(
+				String externalReferenceCode, Long siteId)
+		throws Exception {
+
+		MBMessage mbMessage = _getMbMessageByExternalReferenceCode(
```

Comment:
Here I use the localService (inside the method) because we call again the getMBMessage in the MessageBoardMessageDTOConverter, and there we will check the permissions. This is redundant in some cases, as getSiteMessageBoardMessageByFriendlyUrlPath, but in this case it helps me to not add the method to check permissions in the Service Layer.

Related with this I've found we do this a lot in many DTOConverters (get the entities twice, outside and inside DTOConverters). Although I think there is a database cache, maybe we can give it a look to use `toDTO(DTOConverterContext dtoConverterContext, E object)` to only get the entity once. When we do this we should revisit this code to check for permission when getting the MBMessage by externalReferenceCode.

Reply:
Yes! I added `toDTO(DTOConverterContext dtoConverterContext, E object)` just because that reason, because we had double retrievals everywhere. I changed some calls when we did the flamegraph optimization for Questions.